File: frontend/src/types/ai-model.ts
/**
 * AI Model resource types — mirrors `/api/v1/admin/ai-models` shapes from
 * docs/ai-models-and-embedders-design.md §3 + §7.
 *
 * Security invariants enforced by the API:
 *  - GET endpoints never return plaintext API keys; only `api_key_last4` and `api_key_set`.
 *  - PATCH preserves the existing key when `api_key` is omitted/null.
 */

export type TestConnectionStatus = 'ok' | 'failed' | 'never'

export interface AIModelCapabilities {
  function_calling?: boolean
  vision?: boolean
  json_mode?: boolean
  streaming?: boolean
  max_context_tokens?: number
  /** USD per 1M input tokens — used for sort ordering in pickers. */
  input_cost_per_1m?: number
  /** USD per 1M output tokens. */
  output_cost_per_1m?: number
}

export interface AIModelPublic {
  id: string
  name: string
  provider: string
  base_url: string | null
  model_id: string
  /** Last 4 chars of the encrypted API key, for masked display. */
  api_key_last4: string | null
  api_key_set: boolean
  extra_config: Record<string, unknown>
  default_temperature: number
  default_max_tokens: number
  capabilities: AIModelCapabilities
  is_active: boolean
  last_test_at: string | null
  last_test_status: TestConnectionStatus
  last_test_error: string | null
  created_at: string
  updated_at: string
  created_by: string | null
}

export interface AIModelListResponse {
  items: readonly AIModelPublic[]
  total: number
  limit: number
  offset: number
}

export interface AIModelCreateRequest {
  name: string
  provider: string
  model_id: string
  api_key: string
  base_url?: string | null
  extra_config?: Record<string, unknown>
  default_temperature?: number
  default_max_tokens?: number
  capabilities?: AIModelCapabilities
  is_active?: boolean
}

export interface AIModelUpdateRequest {
  name?: string
  provider?: string
  model_id?: string
  /** Omit/null to preserve existing key. */
  api_key?: string | null
  base_url?: string | null
  extra_config?: Record<string, unknown>
  default_temperature?: number
  default_max_tokens?: number
  capabilities?: AIModelCapabilities
  is_active?: boolean
}

/**
 * Plaintext form-driven test (`POST /test-connection`). Works pre-save and
 * never persists the supplied key.
 */
export interface AIModelTestPlaintextRequest {
  provider: string
  model_id: string
  api_key: string
  base_url?: string | null
  extra_config?: Record<string, unknown>
}

export interface TestConnectionResponse {
  ok: boolean
  latency_ms: number | null
  error: string | null
}

export interface AIModelUsageStage {
  stage: string
  label: string
}

export interface AIModelUsage {
  stages: readonly AIModelUsageStage[]
  /** Number of chat messages whose stage referenced this model. */
  chat_messages_count: number
  total_references: number
}

export interface AIModelDeleteConflict {
  /** Resources that reference this AI Model — admin must reassign first. */
  referenced_by: readonly {
    resource_type: string
    resource_id: string
    label: string
  }[]
}

File: backend/src/schemas/ai_model.py
"""Pydantic schemas for the AIModel admin API.

Public responses NEVER include the plaintext API key — only ``api_key_last4``
and ``api_key_set`` for UX hints.
"""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field


class AIModelCreate(BaseModel):
    """Body for ``POST /api/v1/admin/ai-models``."""

    name: str = Field(..., min_length=1, max_length=150)
    provider: str = Field(..., min_length=1, max_length=64)
    model_id: str = Field(..., min_length=1, max_length=200)
    base_url: str | None = Field(default=None, max_length=500)
    api_key: str | None = Field(default=None, max_length=4000)
    extra_config: dict[str, Any] = Field(default_factory=dict)
    default_temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    default_max_tokens: int = Field(default=2048, gt=0)
    capabilities: dict[str, Any] = Field(default_factory=dict)
    is_active: bool = True


class AIModelUpdate(BaseModel):
    """Body for ``PATCH /api/v1/admin/ai-models/{id}``.

    ``api_key`` is tri-state: omitted/``None`` → preserve existing; non-empty
    string → replace.  An explicit empty string clears the credential.
    """

    name: str | None = Field(default=None, min_length=1, max_length=150)
    provider: str | None = Field(default=None, min_length=1, max_length=64)
    model_id: str | None = Field(default=None, min_length=1, max_length=200)
    base_url: str | None = Field(default=None, max_length=500)
    api_key: str | None = Field(default=None, max_length=4000)
    extra_config: dict[str, Any] | None = None
    default_temperature: float | None = Field(default=None, ge=0.0, le=2.0)
    default_max_tokens: int | None = Field(default=None, gt=0)
    capabilities: dict[str, Any] | None = None
    is_active: bool | None = None

    model_config = ConfigDict(extra="forbid")


class AIModelPublic(BaseModel):
    """GET response model — no plaintext key, ever."""

    id: uuid.UUID
    name: str
    provider: str
    model_id: str
    base_url: str | None
    extra_config: dict[str, Any]
    default_temperature: float
    default_max_tokens: int
    capabilities: dict[str, Any]
    is_active: bool
    api_key_set: bool
    api_key_last4: str | None
    last_test_at: datetime | None
    last_test_status: str | None
    last_test_error: str | None
    created_at: datetime
    updated_at: datetime
    created_by: uuid.UUID | None


class AIModelList(BaseModel):
    items: list[AIModelPublic]
    total: int
    limit: int
    offset: int


class TestConnectionPlaintextRequest(BaseModel):
    """Body for ``POST /ai-models/test-connection`` (does not persist)."""

    provider: str
    model_id: str
    api_key: str
    base_url: str | None = None
    extra_config: dict[str, Any] = Field(default_factory=dict)


class TestConnectionResult(BaseModel):
    ok: bool
    latency_ms: int
    error: str | None = None


class AIModelUsage(BaseModel):
    """Response for ``GET /ai-models/{id}/usage``."""

    stages: list[str]
    chat_messages_count: int


class AIModelDeleteConflict(BaseModel):
    error: Literal["referenced"] = "referenced"
    referenced_by: dict[str, Any]
